src/components/InngestFunction.ts

```diff
@@ -251,288 +252,310 @@ export class InngestFunction<
+            `Function "${this.name}" received a failure event to handle, but no failure handler was defined.`
+          );
+        }
+
+        userFnToRun = this.#onFailureFn;
 
-      if (!this.#onFailureFn) {
-        // TODO PrettyError
-        throw new Error(
-          `Function "${this.name}" received a failure event to handle, but no failure handler was defined.`
+        (fnArg as FailureEventArgs).error = deserializeError(
+          (fnArg.event as FailureEventPayload).data.error
         );
       }
 
-      userFnToRun = this.#onFailureFn;
+      /**
+       * If the user has passed functions they wish to use in their step, add them
+       * here.
+       *
+       * We simply place a thin `tools.run()` wrapper around the function and
+       * nothing else.
+       */
+      if (this.#opts.fns) {
+        fnArg.fns = Object.entries(this.#opts.fns).reduce((acc, [key, fn]) => {
+          if (typeof fn !== "function") {
+            return acc;
+          }
 
-      (fnArg as FailureEventArgs).error = deserializeError(
-        (fnArg.event as FailureEventPayload).data.error
-      );
-    }
+          return {
+            ...acc,
+            // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-call
+            [key]: (...args: unknown[]) => tools.run(key, () => fn(...args)),
+          };
+        }, {});
+      }
 
-    /**
-     * If the user has passed functions they wish to use in their step, add them
-     * here.
-     *
-     * We simply place a thin `tools.run()` wrapper around the function and
-     * nothing else.
-     */
-    if (this.#opts.fns) {
-      fnArg.fns = Object.entries(this.#opts.fns).reduce((acc, [key, fn]) => {
-        if (typeof fn !== "function") {
-          return acc;
+      // eslint-disable-next-line @typescript-eslint/no-misused-promises, no-async-promise-executor
+      const userFnPromise = new Promise(async (resolve, reject) => {
+        if (opStack.length == 0 && !requestedRunStep) {
+          logger.enable();
         }
 
-        return {
-          ...acc,
-          // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-call
-          [key]: (...args: unknown[]) => tools.run(key, () => fn(...args)),
-        };
-      }, {});
-    }
-
-    // eslint-disable-next-line @typescript-eslint/no-misused-promises, no-async-promise-executor
-    const userFnPromise = new Promise(async (resolve, reject) => {
-      try {
-        // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        resolve(await userFnToRun(fnArg as Context<any, any, any>));
-      } catch (err) {
-        reject(err);
-      }
-    });
+        try {
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          resolve(await userFnToRun(fnArg as Context<any, any, any>));
+        } catch (err) {
+          reject(err);
+          logger.error(err);
+        }
+        logger.disable();
```

Comment:
Did you find these were still needed?

Reply:
I found no differences when it comes to output, but decided to keep it for people reading it later.
It's going to be confusing regardless but only seeing `logger.enable` and it never gets disabled seems like it'll pretty hard to comprehend.

basically when something was opened, let's just close it just in case.